framework/src/utils/ADUtils.C

```diff
@@ -0,0 +1,53 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADUtils.h"
+
+#include "SystemBase.h"
+#include "libmesh/system.h"
+#include "libmesh/dof_map.h"
+
+namespace Moose
+{
+
+std::unordered_map<dof_id_type, Real>
+globalDofIndexToDerivative(const ADReal & ad_real,
+                           const SystemBase & sys,
+                           const Elem * elem,
+                           const ElementType elem_type /*=ElementType::Element*/)
+{
+  std::unordered_map<dof_id_type, Real> ret_val;
+
+  const System & libmesh_sys = sys.system();
+  const DofMap & dof_map = libmesh_sys.get_dof_map();
+
+  const unsigned int num_vars = libmesh_sys.n_vars();
+
+  const auto max_dofs_per_elem = sys.getMaxVarNDofsPerElem();
+
+  for (unsigned int var_num = 0; var_num < num_vars; ++var_num)
+  {
+    std::vector<dof_id_type> global_indices;
+
+    // Get the global indices corresponding to var_num that exist on elem
+    dof_map.dof_indices(elem, global_indices, var_num);
+
+    // determine the AD offset for the current var
+    const auto ad_offset = adOffset(var_num, max_dofs_per_elem, elem_type, num_vars);
```

Comment:
What happens if `elem` isn't currently assembly's current elem or neighbor?  Then we just get nonsense right?  Maybe we should guard against that.

Reply:
Or at least we need to know which dof indices were used when the dualnumbers that contributed to `ad_real` were seeded/initialized.

Reply:
Might as well just remove `elem` from the argument list then and just take it from assembly

Reply:
I went ahead and removed `elem` from the argument list, and am just using the assembly elem APIs to get the current elements